apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts

```diff
@@ -52,22 +100,84 @@ router.post(
       throw new error.HttpStatusError(403, 'Not implemented (feature not available)');
     }
+
     const client = new OpenAI({
       apiKey: config.openAiApiKey,
       organization: config.openAiOrganization,
     });
 
-    if (req.body.__action === 'generate_question') {
-      const jobSequenceId = await generateQuestion(
+    if (req.body.__action === 'ai_generate') {
+      const jobSeq = await generateQuestion(
         client,
         res.locals.course ? res.locals.course.id : undefined,
         res.locals.authn_user.user_id,
         req.body.prompt,
       );
-      res.redirect(res.locals.urlPrefix + '/jobSequence/' + jobSequenceId);
+
+      const data = jobSeq.jobData;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, jobSeq.jobSequenceId, res.locals),
+        );
+      } else {
+        throw new error.HttpStatusError(500, `Job Sequence ${jobSeq.jobSequenceId} failed.`);
+      }
     } else if (req.body.__action === 'sync_context_documents') {
       const jobSequenceId = await syncContextDocuments(client, res.locals.authn_user.user_id);
       res.redirect('/pl/jobSequence/' + jobSequenceId);
+    } else if (req.body.__action === 'ai_regen') {
+      const genJobs = await selectJobsByJobSequenceId(req.body.unsafe_sequence_job_id);
+      if (
+        genJobs.length !== 1 ||
+        !genJobs[0]?.course_id ||
+        !idsEqual(genJobs[0]?.course_id, res.locals.course.id)
+      ) {
+        throw new error.HttpStatusError(
+          403,
+          `Course for job sequence ${req.body.unsafe_sequence_job_id} does not match the current course.`,
+        );
+      }
+
+      const jobSeq = await regenerateQuestion(
+        client,
+        res.locals?.course?.course_id,
+        res.locals.authn_user.user_id,
+        genJobs[0]?.data?.prompt,
+        req.body.prompt,
+        genJobs[0]?.data?.html,
+        genJobs[0]?.data?.python,
+      );
+
+      const data = jobSeq.jobData;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, jobSeq.jobSequenceId, res.locals),
+        );
+      } else {
+        res.redirect('/pl/jobSequence/' + jobSeq.jobSequenceId);
+      }
+    } else if (req.body.__action === 'save_question') {
+      const genJobs = await selectJobsByJobSequenceId(req.body.unsafe_sequence_job_id);
+      if (
+        genJobs.length !== 1 ||
+        !genJobs[0]?.course_id ||
+        !idsEqual(genJobs[0]?.course_id, res.locals.course.id)
+      ) {
+        throw new error.HttpStatusError(
+          403,
+          `Course for job sequence ${req.body.unsafe_sequence_job_id} does not match the current course.`,
+        );
+      }
+
+      const qid = await saveGeneratedQuestion(
+        res,
+        genJobs[0]?.data?.html,
+        genJobs[0]?.data?.python,
+      );
+
+      if (qid) {
+        res.redirect(res.locals.urlPrefix + '/question/' + qid + '/settings');
+      }
```

Comment:
What happens if `qid` isn't defined? The request will hang. We should handle this case somehow.

Reply:
it actually wouldn't because the qid-generating function would redirect, but that was too confusing and bad code anyways. changed to something more reasonable in e0331beaf9d587990dcda087d800feedc786fc87.